src/redis/windowed_cache_impl.go

```diff
@@ -0,0 +1,168 @@
+package redis
+
+import (
+	"math/rand"
+
+	"github.com/coocood/freecache"
+	pb "github.com/envoyproxy/go-control-plane/envoy/service/ratelimit/v3"
+	"github.com/envoyproxy/ratelimit/src/algorithm"
+	"github.com/envoyproxy/ratelimit/src/config"
+	"github.com/envoyproxy/ratelimit/src/limiter"
+	"github.com/envoyproxy/ratelimit/src/redis/driver"
+	"github.com/envoyproxy/ratelimit/src/utils"
+	logger "github.com/sirupsen/logrus"
+	"golang.org/x/net/context"
+)
+
+// This rolling window limit implemented using Generic Cell Rate Algorithm (GCRA)
+// GCRA works by tracking remaining limit through a time called the “theoretical arrival time” (TAT).
+// Request cost is represented as a multiplier of “emission interval”, which is derived from the duration of equally spread request.
+// TAT is seeded by the current request arrival if not set then add the request costs.
+// Subtract the window duration from TAT to get the time to allow a request
+// Requests are allowed if the time to allow a request is in the past
+// Store the TAT for next process
+// https://blog.ian.stapletoncordas.co/2018/12/understanding-generic-cell-rate-limiting.html
+
+type windowedRateLimitCacheImpl struct {
+	client driver.Client
+	// Optional Client for a dedicated cache of per second limits.
+	// If this client is nil, then the Cache will use the client for all
+	// limits regardless of unit. If this client is not nil, then it
+	// is used for limits that have a SECOND unit.
+	perSecondClient            driver.Client
+	timeSource                 utils.TimeSource
+	jitterRand                 *rand.Rand
+	expirationJitterMaxSeconds int64
+	cacheKeyGenerator          utils.CacheKeyGenerator
+	localCache                 *freecache.Cache
+	nearLimitRatio             float32
+	algorithm                  algorithm.RatelimitAlgorithm
+}
+
+func (this *windowedRateLimitCacheImpl) DoLimit(
+	ctx context.Context,
+	request *pb.RateLimitRequest,
+	limits []*config.RateLimit) []*pb.RateLimitResponse_DescriptorStatus {
+
+	logger.Debugf("starting windowed cache lookup")
```

Comment:
same here, duplicated code across two redis implementations.

Reply:
yes, most of the code is assigning value to a variable, any concern or suggestion to clean them? extract to function that return
- hitsAddend
- cacheKeys
- isOverLimitWithLocalCache
- tats/results?

Reply:
am thinking about followup PR that would merge those implementations at least based on one dimension (backend type or algorithm type)